File: app/lib/domain/entity/kyc/ex_kyc.dart
import 'package:json_annotation/json_annotation.dart';

part 'ex_kyc.g.dart';

@JsonSerializable()
class ExKyc {
  final int? id;
  @JsonKey(name: 'user_id')
  final int userId;
  @JsonKey(name: 'kyc_level')
  final int kycLevel; // 1=标准认证,2=进阶认证(KYC)
  @JsonKey(name: 'country_code')
  final String? countryCode;
  final String? surname;
  @JsonKey(name: 'middle_name')
  final String? middleName;
  final String? name;
  final int? gender; // 1=男,2=女,3=其他
  final String? birthday;
  @JsonKey(name: 'id_type')
  final String? idType; // ID_CARD,PASSPORT,DRIVING_LICENSE,OTHER
  @JsonKey(name: 'id_number')
  final String? idNumber;
  @JsonKey(name: 'id_issue_date')
  final String? idIssueDate;
  @JsonKey(name: 'id_expiry_date')
  final String? idExpiryDate;
  final String? address;
  final double? latitude;
  final double? longitude;
  @JsonKey(name: 'location_accuracy')
  final double? locationAccuracy;
  @JsonKey(name: 'location_time')
  final String? locationTime;
  @JsonKey(name: 'location_address')
  final String? locationAddress;
  @JsonKey(name: 'id_front_image')
  final String? idFrontImage;
  @JsonKey(name: 'id_back_image')
  final String? idBackImage;
  @JsonKey(name: 'id_selfie_image')
  final String? idSelfieImage;
  @JsonKey(name: 'address_proof_image')
  final String? addressProofImage;
  final int status; // 0=待审核,1=审核中,2=已通过,3=已拒绝,4=已取消
  @JsonKey(name: 'ocr_result')
  final Map<String, dynamic>? ocrResult;
  @JsonKey(name: 'ocr_confidence')
  final double? ocrConfidence;
  @JsonKey(name: 'created_at')
  final String? createdAt;
  @JsonKey(name: 'updated_at')
  final String? updatedAt;

  ExKyc({
    this.id,
    required this.userId,
    required this.kycLevel,
    this.countryCode,
    this.surname,
    this.middleName,
    this.name,
    this.gender,
    this.birthday,
    this.idType,
    this.idNumber,
    this.idIssueDate,
    this.idExpiryDate,
    this.address,
    this.latitude,
    this.longitude,
    this.locationAccuracy,
    this.locationTime,
    this.locationAddress,
    this.idFrontImage,
    this.idBackImage,
    this.idSelfieImage,
    this.addressProofImage,
    required this.status,
    this.ocrResult,
    this.ocrConfidence,
    this.createdAt,
    this.updatedAt,
  });

  factory ExKyc.fromJson(Map<String, dynamic> json) => _$ExKycFromJson(json);

  Map<String, dynamic> toJson() => _$ExKycToJson(this);

  /// 获取状态文本
  String get statusText {
    switch (status) {
      case 0:
        return '待审核';
      case 1:
        return '审核中';
      case 2:
        return '已通过';
      case 3:
        return '已拒绝';
      case 4:
        return '已取消';
      default:
        return '未知';
    }
  }

  /// 获取认证等级文本
  String get kycLevelText {
    switch (kycLevel) {
      case 1:
        return '标准身份认证';
      case 2:
        return '进阶身份认证';
      default:
        return '未知等级';
    }
  }

  /// 获取证件类型文本
  String get idTypeText {
    switch (idType) {
      case 'ID_CARD':
        return '身份证';
      case 'PASSPORT':
        return '护照';
      case 'DRIVING_LICENSE':
        return '驾驶证';
      case 'OTHER':
        return '其他';
      default:
        return '未知';
    }
  }

  /// 获取性别文本
  String get genderText {
    switch (gender) {
      case 1:
        return '男';
      case 2:
        return '女';
      case 3:
        return '其他';
      default:
        return '未知';
    }
  }

  /// 获取完整姓名
  String get fullName {
    final parts = <String>[];
    if (surname != null && surname!.isNotEmpty) parts.add(surname!);
    if (middleName != null && middleName!.isNotEmpty) parts.add(middleName!);
    if (name != null && name!.isNotEmpty) parts.add(name!);
    return parts.join(' ');
  }

  /// 获取脱敏后的证件号码
  String get maskedIdNumber {
    if (idNumber == null || idNumber!.length < 6) return idNumber ?? '';
    return '${idNumber!.substring(0, 2)}${'*' * (idNumber!.length - 4)}${idNumber!.substring(idNumber!.length - 2)}';
  }

  /// 是否已通过认证
  bool get isApproved => status == 2;

  /// 是否待审核或审核中
  bool get isPending => status == 0 || status == 1;

  /// 是否已拒绝
  bool get isRejected => status == 3;
}

File: server/plugin/ds/ex/src/Model/ExKyc.php
<?php

declare(strict_types=1);

namespace Plugin\Ds\Ex\Model;

use Hyperf\DbConnection\Model\Model;

/**
 * KYC认证申请表模型
 *
 * @property int $user_id 用户ID
 * @property int $kyc_level 认证等级：1=标准认证,2=进阶认证(KYC)
 * @property string $country_code 国家/地区代码（ISO 3166-1 alpha-2）
 * @property string $surname 姓
 * @property string $middle_name 中间名
 * @property string $name 名字
 * @property int $gender 性别：1=男,2=女,3=其他
 * @property \Carbon\Carbon $birthday 出生日期
 * @property string $id_type 证件类型：ID_CARD=身份证,PASSPORT=护照,DRIVING_LICENSE=驾驶证,OTHER=其他
 * @property string $id_number 身份证件号码（加密存储）
 * @property string $id_number_hash 身份证件号码哈希（用于去重）
 * @property \Carbon\Carbon $id_issue_date 证件签发日期
 * @property \Carbon\Carbon $id_expiry_date 证件有效期
 * @property string $address 地址信息
 * @property float $latitude GPS纬度
 * @property float $longitude GPS经度
 * @property float $location_accuracy 定位精度（米）
 * @property \Carbon\Carbon $location_time 定位时间
 * @property string $location_address GPS反地理编码地址（根据经纬度获取的地址）
 * @property string $id_front_image 证件正面照片URL
 * @property string $id_back_image 证件背面照片URL
 * @property string $id_selfie_image 手持证件自拍照片URL
 * @property string $address_proof_image 地址证明照片URL
 * @property int $status 审核状态：0=待审核,1=已通过,2=已拒绝,3=已取消
 * @property array $ocr_result OCR识别结果：{&quot;id_number&quot;:&quot;&quot;,&quot;name&quot;:&quot;&quot;,&quot;birthday&quot;:&quot;&quot;,&quot;address&quot;:&quot;&quot;,...}
 * @property float $ocr_confidence OCR识别置信度（0-100）
 * @property string $remark 备注,拒绝原因
 * @property \Carbon\Carbon $created_at 创建时间
 * @property \Carbon\Carbon $updated_at 更新时间
 */
class ExKyc extends Model
{
    protected ?string $table = 'ex_kyc';

    protected array $fillable = [
        'user_id',
        'kyc_level',
        'country_code',
        'surname',
        'middle_name',
        'name',
        'gender',
        'birthday',
        'id_type',
        'id_number',
        'id_number_hash',
        'id_issue_date',
        'id_expiry_date',
        'address',
        'latitude',
        'longitude',
        'location_accuracy',
        'location_time',
        'location_address',
        'id_front_image',
        'id_back_image',
        'id_selfie_image',
        'address_proof_image',
        'status',
        'ocr_result',
        'ocr_confidence',
        'remark',
        'created_at',
        'updated_at',
    ];

    protected array $casts = [
        'user_id' => 'integer',
        'kyc_level' => 'integer',
        'country_code' => 'string',
        'surname' => 'string',
        'middle_name' => 'string',
        'name' => 'string',
        'gender' => 'integer',
        'birthday' => 'date',
        'id_type' => 'string',
        'id_number' => 'string',
        'id_number_hash' => 'string',
        'id_issue_date' => 'date',
        'id_expiry_date' => 'date',
        'address' => 'string',
        'latitude' => 'decimal:2',
        'longitude' => 'decimal:2',
        'location_accuracy' => 'decimal:2',
        'location_time' => 'datetime',
        'location_address' => 'string',
        'id_front_image' => 'string',
        'id_back_image' => 'string',
        'id_selfie_image' => 'string',
        'address_proof_image' => 'string',
        'status' => 'integer',
        'ocr_result' => 'array',
        'ocr_confidence' => 'decimal:2',
        'remark' => 'string',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];
    protected array $hidden = [];


    public function isKyc(): int
    {
        //1标准认证审核中 2标准认证完成 3标准认证未通过 4进阶认证审核中 5进阶认证完成 6进阶认证未通过
        if ($this->kyc_level == 2) return $this->status == 2 ? 6 : ($this->status ? 5 : 4);
        return $this->status == 2 ? 3 : ($this->status ? 2 : 1);
    }
}
